draft-ietf-quic-tls.md

```diff
@@ -385,13 +385,9 @@ perspective of the endpoint in question.
 
 ### Handshake Confirmed {#handshake-confirmed}
 
-In this document, the TLS handshake is considered confirmed at an endpoint when
-the following two conditions are met: the handshake is complete, and the
-endpoint has received an acknowledgment for a packet sent with 1-RTT keys.
-This second condition can be implemented by recording the lowest packet number
-sent with 1-RTT keys, and the highest value of the Largest Acknowledged field
-in any received 1-RTT ACK frame: once the latter is higher than or equal to the
-former, the handshake is confirmed.
+In this document, the TLS handshake is considered confirmed at the server when
+the handshake completes. At the client, the handshake is considered confirmed
+when the HANDSHAKE_DONE frame is received.
```

Comment:
I don't think that we should add an "OR" here.  If the server acts badly, then the handshake won't ever be confirmed and that's OK.

Reply:
If we're going to add a new signal, I'd like to enforce that.  And if the rule is that the server must send HANDSHAKE_DONE in a short period of time, then it's reasonable for the client to assume any ACK in 1-RTT prior to receiving the HANDSHAKE_DONE frame is a result of the packet containing HANDSHAKE_DONE being lost.

I don't see a downside of this addition, given it's quite simple.

Reply:
Overall, I quite like this HANDSHAKE_DONE proposal. It seems obviously correct which, given our previous iterations is rather appealing :)

I agree with Ian that allowing the client to drop handshake keys when it receives an ACK of a 1-RTT packet sounds like a great optimization which helps ensure the keys get dropped quickly.

Reply:
I guess the only danger is that putting in this backstop makes it semi-OK for servers to not bother to send HANDSHAKE_DONE as an optimization, and then all the client data goes in 0RTT packets and we end up in the deadlock.

I would rather noncompliance result in holding keys, rather than good connections timing out.

Reply:
The server could also not send anything in Handshake and force the client to continue using 0RTT packets, which seems equivalent to what you're describing?  If so, then this is nothing new.

Maybe we need more motivating text about why you really need to send HANDSHAKE_DONE, but I think this design is generally quite efficient.

If we REALLY want to ensure the client gets the signal, we can add a key phase flip between 0.5RTT and 1RTT as well, but then do we change keys as well?

Reply:
@ianswett That's definitely not what I was talking about. But I see now that my example is not right.